tools/azsdk-cli/Azure.Sdk.Tools.Cli/Tools/Package/SampleGeneration/DotNetSourceInputProvider.cs

```diff
@@ -0,0 +1,167 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System.Collections.Concurrent;
+using Azure.Sdk.Tools.Cli.Helpers;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+
+namespace Azure.Sdk.Tools.Cli.SampleGeneration;
+
+/// <summary>
+/// Discovers C# source and supporting infrastructure files for .NET client libraries.
+/// </summary>
+public class DotNetSourceInputProvider : ILanguageSourceInputProvider
+{
+
+    private static readonly string[] s_testFileNameIndicators =
+    [
+        "Tests.cs",
+        "Test.cs",
+        "Facts.cs",
+        "Spec.cs",
+    ];
+
+    private static readonly HashSet<string> s_testMethodAttributes = new(StringComparer.OrdinalIgnoreCase)
+    { "Test", "Fact", "Theory", "TestCase", "DataTestMethod", "Repeat" };
+
+    private static readonly HashSet<string> s_lifecycleAttributes = new(StringComparer.OrdinalIgnoreCase)
+    { "SetUp", "TearDown", "OneTimeSetUp", "OneTimeTearDown", "ClassInitialize", "ClassCleanup", "AssemblyInitialize", "AssemblyCleanup" };
+
+    private static readonly HashSet<string> s_infraBaseTypeIndicators = new(StringComparer.OrdinalIgnoreCase)
+    { "RecordedTestBase", "ClientTestBase", "SamplesBase" };
+
+    private static readonly ConcurrentDictionary<string, (DateTime LastWrite, bool IsTest)> s_cache = new();
+
+    public IReadOnlyList<FileHelper.SourceInput> Create(string packagePath)
+    {
+        var result = new List<FileHelper.SourceInput>();
+
+        var src = Path.Combine(packagePath, "src");
+        if (Directory.Exists(src))
+        {
+            result.Add(new FileHelper.SourceInput(src, IncludeExtensions: [".cs"]));
+        }
+
+        var tests = Path.Combine(packagePath, "tests");
+        if (Directory.Exists(tests))
+        {
+            foreach (var file in EnumerateInfrastructureFiles(tests))
+            {
+                result.Add(new FileHelper.SourceInput(file));
+            }
+        }
+
+        var samples = Path.Combine(packagePath, "tests", "samples");
+        if (Directory.Exists(samples))
+        {
+            result.Add(new FileHelper.SourceInput(samples, IncludeExtensions: [".cs"]));
+        }
+
+        return result;
+    }
+
+    private static IEnumerable<string> EnumerateInfrastructureFiles(string testsDirectory)
+    {
+        foreach (var file in Directory.EnumerateFiles(testsDirectory, "*.cs", SearchOption.AllDirectories))
+        {
+            if (!IsTestFile(file))
+            {
+                yield return file;
+            }
+        }
+    }
+
+    private static bool IsTestFile(string filePath)
```

Comment:
Are we really saying here that the best way to check if a file is a test file is

- Check if the file name ends in one of those suffixes (what if we have a source file which ends in Test for some reason? Seems plausible)
- If the file name does not match one of those suffixes, parse it and check for common test attributes (Test, Fact, Theory...) (what test library are we using in the .NET SDK repo? Surely we can narrow this down, or are we mixing NUnit/MSTest/Xunit in different places)
  - If parsing fails then we try to check for attributes _manually_ by text search?

Is there really no better way to do this?